### PR TITLE
basic authentication now follows redirects!

### PR DESCRIPTION
**Pull request by nerdmaennchen**
this should fix https://github.com/nextcloud/user_external/issues/45

all the best,
Lutz

### CODE REVIEW

File: lib/basicauth.php [hunk 1]
```diff
@@ -39,13 +39,21 @@ public function checkPassword($uid, $password) {
...
-		$returnCode= substr($headers[0], 9, 3);
-		if(substr($returnCode, 0, 1) === '2') {
-			$this->storeUser($uid);
-			return $uid;
-		} else {
-			return false;
+		/* get_headers follows redirects up to a maximum (default: 20)
+		 * the response code of the last request is stored in the numerically greatest item
+		 */
+		$response = $headers[max(array_keys($headers))];
```
Comment (violoncelloCH) [inline lib/basicauth.php]: what should `max()` do here? take the last header?

**Comment by violoncelloCH** [comment]
ahh, get_headers() should follow redirects by default?

**Comment by nerdmaennchen** [comment]
heyhey,

sorry for my late reply.
I was surprised as well when I found out that get_headers already does follow redirects.
regarding the use of max():
By all means I am no expert on php (all php code I've ever touched is the code you're dealing with just now).
Anyways, after playing with the function for a bit more I figured there are two problems:

It only works if the headers in the response do not have numbers as keys (which they should not have anyways... I consider this a minor issue)
Also, it only works in no-redirect situations if the HTTP-response field containing the error code is the first element.

I'm actually fairly sure that get_headers will always behave in a way that the element with index 0 will be the first of the returned array but the way my code utilizes this assumption is a bit of a stretch.
I'm going to overhaul this part.

Thanks for pointing out the issues that come with max()!

Greetings from Guatemala,
a lovely place with lovely people!

**Comment by nerdmaennchen** [comment]
I think using a regex for response parsing would be a bit overkill.
However, I think looking for the numerically greatest element of all elements with integer keys should do the trick.
My reasoning is that header fields are a mapping from string to string and checking the type of the array indexes should suffice.

File: lib/basicauth.php [hunk 2]
```diff
@@ -39,13 +39,27 @@ public function checkPassword($uid, $password) {
...
-		if(substr($returnCode, 0, 1) === '2') {
-			$this->storeUser($uid);
-			return $uid;
-		} else {
-			return false;
+		/* get_headers() follows redirects up to a maximum (default: 20)
+		 * the response code of the last request is stored in the numerically greatest item
+		 * $headers[0] is always present
+		 */
+		$response_idx = 0;
```
Comment (violoncelloCH) [inline lib/basicauth.php]: Nextcloud code styling prefers using `CamelCase` over `snake_case`.

**Pull request merged**